### ISSUE TITLE
[Accepted] SDL 0041 - Provide AppIcon resumption across app registration requests 

### ISSUE THREAD

**Issue by theresalech** [smartdevicelink/sdl_evolution#127]
Hello SDL community,

The review of "Provide AppIcon resumption across app registration requests" begins now and runs through March 28, 2017. The proposal is available here:

https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0041-appicon-resumption.md

Reviews are an important part of the SDL evolution process. All reviews should be sent to the associated Github issue at:

https://github.com/smartdevicelink/sdl_evolution/issues/127

What goes into a review?

The goal of the review process is to improve the proposal under review through constructive criticism and, eventually, determine the direction of SDL. When writing your review, here are some questions you might want to answer in your review:

* Is the problem being addressed significant enough to warrant a change to SDL?
* Does this proposal fit well with the feel and direction of SDL?
* If you have used competitors with a similar feature, how do you feel that this proposal compares to those?
* How much effort did you put into your review? A glance, a quick reading, or an in-depth study?
Please state explicitly whether you believe that the proposal should be accepted into SDL.

More information about the SDL evolution process is available at

https://github.com/smartdevicelink/sdl_evolution/blob/master/process.md

Thank you,
Theresa Lech

Program Manager - Livio
theresa@livio.io

**Comment by stefanek22** [comment]
Agree with the proposal, as it provides a smoother user experience across drive cycles.  Option 2 should also be considered, as the existing "state" of the app could be rolled into the existing hash mechanism.

**Comment by joeljfischer** [comment]
I have a concern about this proposal:

1. The updated RAIR RPC states:

```
<function name="RegisterAppInterface" functionID="RegisterAppInterfaceID" messagetype="response">
<description>The response to registerAppInterface</description>

    <param name="iconResumed" type="Boolean" mandatory="true">
        <description>Existence of apps icon at system. If true, apps icon was resumed at system. If false, apps icon is not resumed at system</description>
    </param>
</function>
```

If this is a new mandatory parameter without a default value, how will that affect backward compatibility? It seems to me that this should not be mandatory, and if it's not set (e.g. in older systems that don't even know it exists) the mobile libraries should go through the normal icon setting process.

#### Conclusions

> Is the problem being addressed significant enough to warrant a change to SDL?

Yes, this seems like a good change in principle.

> Does this proposal fit well with the feel and direction of SDL?

Yes, this seems like a good approach. I don't believe either alternative option 1 or 2 should be considered.

> If you have used competitors with a similar feature, how do you feel that this proposal compares to those?

Having icons appear immediately seems like a positive change.

> How much effort did you put into your review? A glance, a quick reading, or an in-depth study?

A quick reading.

> Please state explicitly whether you believe that the proposal should be accepted into SDL.

I cannot yet support this proposal without the above questions being answered.

**Comment by mrapitis** [comment]
@joeljfischer thanks for the feedback.  I would be open to having the iconResumed parameter as optional as I see no downsides -- hopefully the group can weigh in with their input as well.

**Comment by mfoedisch** [comment]
Here are my concerns with this proposal:

- essentially, the approach would add ANOTHER way of doing something that can be done already:
-- send file and then set the app icon (which can take a moment)
-- use resumption approach (could be instantaneous)
- by providing a bool "iconResumed" to the app will not tell WHICH icon was used
- if the main concern is to have an icon available immediately at the time of registration, other workarounds could be considered, e.g. the HMI could remember what icon was used last for a particular app and use that one until another one is provided by the app
- Option 1: this would also add another way of doing something

- Option 2: as it is suggesting to use the existing resumption approach with hash id, this seems to make most sense. I'm not sure how "supplementing" would work, though. How would the original proposal get mixed with Option 2?

So, in summary, for consistency reasons I'd propose to stick with existing approaches (i.e. resumption/hash id) or leave it to the HMI to implement a workaround.

**Comment by mrapitis** [comment]
@mfoedisch thanks for your input.  If the consensus wanted to pursue option 2, we may be able accept this proposal with modifications and highlight the requirement of a hashid in order to resume the app icon.

**Comment by Toyota-Sbetts** [comment]
In general we agree with mfoedisch's analysis.  The HMI can already remember the last set app icon and update if/when the app set's the new icon.  We also think many HMI's may handle this process very differently for many reasons or even not support dynamic app icons.   Regardless if we add this change or rely on the HMI for a "workaround", we will still have a gap the first time an app registers.

With that being said, we also don't really see any negative impact of this proposal either.  The given potential downside essentially exists either way for first app registration, so if this proposal is beneficial to some HMI implementations then we can support it as an optional parameter.

**Comment by kshala-ford** [comment]
> The mobile app will be notified if an app icon was successfully resumed via a Boolean iconResumed parameter in the RegisterAppInterface Response message from SDL Core.

The RPC parameter is just a hint to the app to skip the app icon procedure. it's not a must have. 

The app developers I know don't like to use resumeHash. Besides app icon it reloads choices and commands. Therefore not a good alternative.

**Comment by mfoedisch** [comment]
The issue described in https://github.com/smartdevicelink/sdl_evolution/issues/128 kind of morphed (from handling erroneous app behavior) to support "post loading" of images.

Could we consider this as an alternative approach to handle the issue described here? RegisterAppInterface could provide an (optional) AppIcon reference. If it exists, it will be used immediately, if not it will be ignored and post-loading might make it available.

This would become an approach consistent with something existing and not add an island solution. Also, providing a specific appIconName in RAI would make it clear which filename should be used.

**Comment by mrapitis** [comment]
@mfoedisch that is a possibility as well.  Especially since we have an existing placeholder parameter inside of the mobile api that could serve this purpose.

https://github.com/smartdevicelink/sdl_core/blob/master/src/components/interfaces/MOBILE_API.xml#L1302-L1304

**Comment by theresalech** [comment]
The Steering Committee determined that they need more time to review this proposal in order to vote.  The Steering Committee will review this proposal in greater detail and vote on the proposal during the next meeting on April 4, 2017.

**Comment by Harshacharan** [comment]
_If the mobile app had not previously set an app icon successfully and / or the file does not exist on the head unit, SDL Core and the HMI will utilize the default app icon._
@mrapitis  
Hello!
Could you explain the meaning of default app icon..
Is it like some default icons are already stored in the head unit and will be displayed on the HMI based on the type of app whenever the custom app icon is not available?

**Comment by mrapitis** [comment]
@Harshacharan Thanks for your feedback.  Correct, the default app icon is the icon set by the system when the mobile has not yet set its own icon, or when an icon is not present for the particular mobile app.  It can be a very generic icon, mostly used as a placeholder until the mobile app properly sets its own custom icon.  I believe @mfoedisch was potentially suggesting the name of the app icon could be sent across via the RegisterAppInterface request which could potentially simplify the set app icon process.

**Comment by theresalech** [comment]
Upon further review, the Steering Committee has agreed to accept this proposal.

**Comment by theresalech** [comment]
Issues entered:
[Core](https://github.com/smartdevicelink/sdl_core/issues/1456)
[iOS](https://github.com/smartdevicelink/sdl_ios/issues/600)
[Android](https://github.com/smartdevicelink/sdl_android/issues/453)
